### ISSUE TITLE
conda-smithy init fails inside existing repo

### ISSUE THREAD

**Issue by shadowwalkersb** [conda-forge/conda-smithy#437]
I don't really remember if `conda smithy init` has ever worked for me. I remember that I have been doing it manually. Today, finally, I have decided to post the traceback.

```
$ conda smithy init some_dir/recipe
Traceback (most recent call last):
  File "/Users/shadow_walker/anaconda/bin/conda-smithy", line 6, in <module>
    sys.exit(conda_smithy.cli.main())
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/conda_smithy/cli.py", line 289, in main
    args.subcommand_func(args)
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/conda_smithy/cli.py", line 116, in __call__
    generate_feedstock_content(feedstock_directory, args.recipe_directory, meta)
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/conda_smithy/cli.py", line 27, in generate_feedstock_content
    configure_feedstock.copytree(source_recipe_dir, target_recipe_dir)
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/conda_smithy/configure_feedstock.py", line 519, in copytree
    copy_file(s, d)
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/conda_smithy/feedstock_io.py", line 87, in copy_file
    repo.index.add([dst])
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/git/index/base.py", line 745, in add
    entries_added.extend(self._entries_for_paths(paths, path_rewriter, fprogress, entries))
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/git/util.py", line 55, in wrapper
    return func(self, *args, **kwargs)
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/git/index/util.py", line 84, in set_git_working_dir
    return func(self, *args, **kwargs)
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/git/index/base.py", line 631, in _entries_for_paths
    entries_added.append(self._store_path(filepath, fprogress))
  File "/Users/shadow_walker/anaconda/lib/python2.7/site-packages/git/index/base.py", line 590, in _store_path
    st = os.lstat(filepath)     # handles non-symlinks as well
OSError: [Errno 2] No such file or directory: './szip-feedstock/recipe/bld.bat'
```

**Comment by jakirkham** [comment]
Could you please provide some more info with this `conda info`, `conda list -n root`, the recipe in question?

Also I think `conda-smithy` has some issues if the directory and recipe name do not match. So might be worth trying that as the traceback suggestions that they do not match. Not that this should be the case, just I do recall issues like this plaguing staged-recipes.

**Comment by shadowwalkersb** [comment]
```
$ conda info
Current conda install:

               platform : osx-64
          conda version : 4.2.13
       conda is private : False
      conda-env version : 4.2.13
    conda-build version : 2.1.1
         python version : 2.7.12.final.0
       requests version : 2.12.4
       root environment : /Users/shadow_walker/anaconda  (writable)
    default environment : /Users/shadow_walker/anaconda
       envs directories : /Users/shadow_walker/anaconda/envs
          package cache : /Users/shadow_walker/anaconda/pkgs
           channel URLs : https://conda.anaconda.org/GorgonCryoEM/label/dev/osx-64
                          https://conda.anaconda.org/GorgonCryoEM/label/dev/noarch
                          https://conda.anaconda.org/GorgonCryoEM/osx-64
                          https://conda.anaconda.org/GorgonCryoEM/noarch
                          https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.continuum.io/pkgs/free/osx-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/osx-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : /Users/shadow_walker/.condarc
           offline mode : False
```

**Comment by shadowwalkersb** [comment]
```
$ conda list -n root
# packages in environment at /Users/shadow_walker/anaconda:
#
_nb_ext_conf              0.3.0                    py27_0    defaults
alabaster                 0.7.9                    py27_0    conda-forge
altgraph                  0.11                     py27_0    erik
anaconda                  custom                   py27_0    defaults
anaconda-client           1.5.1                    py27_0    conda-forge
appnope                   0.1.0                    py27_0    defaults
appscript                 1.0.1                    py27_0    defaults
argcomplete               1.0.0                    py27_1    defaults
astropy                   1.3                 np111py27_0    defaults
babel                     2.3.4                    py27_0    conda-forge
backports                 1.0                      py27_0    defaults
backports.shutil_get_terminal_size 1.0.0                    py27_0    conda-forge
backports_abc             0.5                      py27_0    conda-forge
beautifulsoup4            4.5.1                    py27_0    defaults
bitarray                  0.8.1                    py27_0    defaults
blas                      1.1                    openblas    conda-forge
bokeh                     0.12.3                   py27_0    defaults
boost                     1.63.0                   py27_2    conda-forge
boto                      2.45.0                   py27_0    defaults
boto3                     1.4.2                    py27_0    conda-forge
botocore                  1.4.81                   py27_0    conda-forge
bottleneck                1.2.0               np111py27_0    conda-forge
bsddb                     1.0                      py27_1    defaults
bzip2                     1.0.6                         3    defaults
ca-certificates           2016.9.26                     0    conda-forge
cdecimal                  2.3                      py27_2    defaults
certifi                   2016.9.26                py27_0    conda-forge
cffi                      1.9.1                    py27_0    defaults
chardet                   2.3.0                    py27_0    defaults
chest                     0.2.3                    py27_0    defaults
click                     6.6                      py27_1    conda-forge
cloudpickle               0.2.1                    py27_2    conda-forge
clyent                    1.2.1                    py27_0    conda-forge
colorama                  0.3.7                    py27_0    conda-forge
conda                     4.3.6                    py27_0    defaults
conda-build               2.1.1                    py27_0    defaults
conda-build-all           1.0.1                    py27_1    conda-forge
conda-env                 2.6.0                         0    conda-forge
conda-manager             0.4.0                    py27_0    defaults
conda-smithy              2.0.0                    py27_0    shadow_walker/label/dev
conda-smithy-2.0.1.dev4   develop                   <pip>
conda-verify              2.0.0                    py27_0    defaults
configobj                 5.0.6                    py27_0    defaults
configparser              3.5.0                    py27_0    conda-forge
contextlib2               0.5.4                    py27_0    conda-forge
cryptography              1.7.1                    py27_0    defaults
curl                      7.49.1                        2    conda-forge
cycler                    0.10.0                   py27_0    conda-forge
cython                    0.25.2                   py27_0    conda-forge
cytoolz                   0.8.2                    py27_0    conda-forge
dask                      0.12.0                   py27_0    conda-forge
datashape                 0.5.4                    py27_0    defaults
db                        5.3.28                        1    defaults
decorator                 4.0.10                   py27_0    conda-forge
dill                      0.2.5                    py27_1    conda-forge
distributed               1.14.3                   py27_0    conda-forge
docutils                  0.12                     py27_0    conda-forge
dynd-python               0.7.2                    py27_0    defaults
entrypoints               0.2.2                    py27_0    conda-forge
enum34                    1.1.6                    py27_1    conda-forge
et_xmlfile                1.0.1                    py27_0    defaults
expat                     2.1.0                         2    conda-forge
fastcache                 1.0.2                    py27_1    defaults
fftw                      3.3.5                         0    conda-forge
filelock                  2.0.6                    py27_0    conda-forge
flask                     0.11.1                   py27_0    conda-forge
flask-cors                3.0.2                    py27_0    conda-forge
freetype                  2.6.3                         1    conda-forge
ftgl                      2.1.3                         0    local
funcsigs                  1.0.2                    py27_0    conda-forge
functools32               3.2.3.2                  py27_1    conda-forge
futures                   3.0.5                    py27_0    conda-forge
get_terminal_size         1.0.0                    py27_0    defaults
gevent                    1.2.0                    py27_0    defaults
gitdb                     0.6.4                    py27_0    conda-forge
gitpython                 2.0.8                    py27_1    conda-forge
gmp                       6.1.1                         1    conda-forge
gmpy2                     2.0.7                    py27_1    conda-forge
greenlet                  0.4.11                   py27_0    defaults
grin                      1.2.1                    py27_3    defaults
gsl                       2.2.1                         0    conda-forge
hdf5                      1.8.17                        8    conda-forge
heapdict                  1.0.0                    py27_0    conda-forge
icu                       56.1                          4    conda-forge
idna                      2.1                      py27_0    conda-forge
imagesize                 0.7.1                    py27_0    conda-forge
ipaddress                 1.0.16                   py27_0    conda-forge
ipykernel                 4.5.2                    py27_0    conda-forge
ipython                   5.1.0                    py27_1    conda-forge
ipython_genutils          0.1.0                    py27_0    conda-forge
ipywidgets                5.2.2                    py27_2    conda-forge
itsdangerous              0.24                     py27_0    conda-forge
jbig                      2.1                           0    defaults
jdcal                     1.3                      py27_0    defaults
jedi                      0.9.0                    py27_0    conda-forge
jinja2                    2.8                      py27_1    conda-forge
jmespath                  0.9.0                    py27_0    conda-forge
jpeg                      9b                            0    conda-forge
jsonschema                2.5.1                    py27_0    conda-forge
jupyter_client            4.4.0                    py27_0    conda-forge
jupyter_console           5.0.0                    py27_0    conda-forge
jupyter_core              4.2.1                    py27_0    conda-forge
libdynd                   0.7.2                         0    defaults
libgcc                    4.8.5                         1    defaults
libgfortran               3.0.0                         0    conda-forge
libiconv                  1.14                          4    conda-forge
libpng                    1.6.26                        0    conda-forge
libsodium                 1.0.10                        0    conda-forge
libtiff                   4.0.6                         7    conda-forge
libxml2                   2.9.4                         3    conda-forge
libxslt                   1.1.29                        2    conda-forge
llvmlite                  0.15.0                   py27_0    defaults
locket                    0.2.0                    py27_1    conda-forge
lxml                      3.7.1                    py27_0    conda-forge
macholib                  1.6                      py27_0    erik
markupsafe                0.23                     py27_0    conda-forge
matplotlib                1.5.3               np111py27_4    conda-forge
mistune                   0.7.3                    py27_0    conda-forge
mkl                       11.3.3                        0    defaults
mkl-service               1.1.2                    py27_2    defaults
mock                      2.0.0                    py27_0    conda-forge
modulegraph               0.11.1                   py27_0    erik
mpc                       1.0.3                         3    conda-forge
mpfr                      3.1.4                         3    conda-forge
mpi4py                    2.0.0                    py27_1    conda-forge
mpich                     3.2                           3    conda-forge
mpmath                    0.19                     py27_1    conda-forge
msgpack-python            0.4.8                    py27_0    conda-forge
multipledispatch          0.4.8                    py27_0    conda-forge
nb_anacondacloud          1.2.0                    py27_0    conda-forge
nb_conda                  2.0.0                    py27_0    conda-forge
nb_conda_kernels          2.0.0                    py27_0    conda-forge
nbconvert                 4.2.0                    py27_0    conda-forge
nbformat                  4.2.0                    py27_0    conda-forge
nbpresent                 3.0.2                    py27_1    conda-forge
ncurses                   5.9                          10    conda-forge
networkx                  1.11                     py27_0    conda-forge
nltk                      3.2.1                    py27_0    conda-forge
nose                      1.3.7                    py27_2    conda-forge
notebook                  4.3.1                    py27_0    conda-forge
numba                     0.30.0              np111py27_0    defaults
numexpr                   2.6.1               np111py27_0    conda-forge
numpy                     1.11.3          py27_blas_openblas_200  [blas_openblas]  conda-forge
odo                       0.5.0                    py27_1    defaults
openblas                  0.2.19                        0    conda-forge
openmpi                   2.0.1                         1    conda-forge
openpyxl                  2.4.0                    py27_0    defaults
openssl                   1.0.2h                        3    conda-forge
pandas                    0.19.2              np111py27_1    conda-forge
partd                     0.3.7                    py27_0    conda-forge
path.py                   8.2.1                    py27_0    conda-forge
pathlib2                  2.1.0                    py27_0    conda-forge
patsy                     0.4.1                    py27_0    conda-forge
pbr                       1.10.0                   py27_0    conda-forge
pep8                      1.7.0                    py27_0    conda-forge
pexpect                   4.2.0                    py27_1    conda-forge
pickleshare               0.7.3                    py27_0    conda-forge
pip                       9.0.1                    py27_0    conda-forge
pkginfo                   1.2.1                    py27_0    conda-forge
ply                       3.9                      py27_0    defaults
prompt_toolkit            1.0.9                    py27_0    conda-forge
psutil                    4.3.0                    py27_0    conda-forge
ptyprocess                0.5.1                    py27_0    conda-forge
py                        1.4.31                   py27_0    conda-forge
py2app                    0.8                      py27_0    erik
pyasn1                    0.1.9                    py27_0    conda-forge
pyaudio                   0.2.7                    py27_0    defaults
pycosat                   0.6.1                    py27_0    conda-forge
pycparser                 2.17                     py27_0    conda-forge
pycrypto                  2.6.1                    py27_0    conda-forge
pycurl                    7.43.0                   py27_0    defaults
pyflakes                  1.3.0                    py27_0    conda-forge
pygithub                  1.29                     py27_0    conda-forge
pygments                  2.1.3                    py27_1    conda-forge
pyopengl                  3.1.0                    py27_0    conda-forge
pyopenssl                 16.2.0                   py27_0    defaults
pyparsing                 2.1.10                   py27_0    conda-forge
pyqt                      4.11.4                   py27_2    conda-forge
pytest                    3.0.5                    py27_0    conda-forge
python                    2.7.12                        1    conda-forge
python-dateutil           2.6.0                    py27_0    conda-forge
python.app                1.2                      py27_4    defaults
pytz                      2016.10                  py27_0    conda-forge
pyyaml                    3.11                     py27_0    conda-forge
pyzmq                     16.0.2                   py27_0    conda-forge
qt                        4.8.7                         4    defaults
qtawesome                 0.3.3                    py27_0    conda-forge
qtpy                      1.1.1                    py27_0    conda-forge
readline                  6.2                           0    conda-forge
redis                     3.2.0                         0    defaults
redis-py                  2.10.5                   py27_0    conda-forge
requests                  2.12.4                   py27_0    conda-forge
rope                      0.9.4                    py27_1    defaults
ruamel.ordereddict        0.4.6                    py27_0    conda-forge
ruamel.yaml               0.11.11                  py27_1    conda-forge
ruamel_yaml               0.11.14                  py27_0    conda-forge
s3fs                      0.0.8                    py27_0    conda-forge
s3transfer                0.1.9                    py27_0    conda-forge
scikit-learn              0.18            np111py27_blas_openblas_203  [blas_openblas]  conda-forge
scipy                     0.18.1          np111py27_blas_openblas_201  [blas_openblas]  conda-forge
setuptools                32.3.0                   py27_0    conda-forge
simplegeneric             0.8.1                    py27_0    conda-forge
singledispatch            3.4.0.3                  py27_0    conda-forge
sip                       4.18                     py27_0    conda-forge
six                       1.10.0                   py27_1    conda-forge
smmap                     0.9.0                    py27_0    conda-forge
snowballstemmer           1.2.1                    py27_0    conda-forge
sockjs-tornado            1.0.3                    py27_0    defaults
sphinx                    1.5                      py27_0    conda-forge
sphinx_rtd_theme          0.1.9                    py27_0    conda-forge
sqlalchemy                1.1.4                    py27_0    defaults
sqlite                    3.13.0                        1    conda-forge
ssl_match_hostname        3.5.0.1                  py27_0    conda-forge
statsmodels               0.6.1               np111py27_0    conda-forge
sympy                     1.0                      py27_0    conda-forge
tblib                     1.3.0                    py27_0    defaults
termcolor                 1.1.0                    py27_0    conda-forge
terminado                 0.6                      py27_0    conda-forge
tk                        8.5.19                        1    conda-forge
toolz                     0.8.2                    py27_0    conda-forge
tornado                   4.4.2                    py27_0    conda-forge
traitlets                 4.3.1                    py27_0    conda-forge
unicodecsv                0.14.1                   py27_0    defaults
wcwidth                   0.1.7                    py27_0    conda-forge
werkzeug                  0.11.10                  py27_0    conda-forge
wheel                     0.29.0                   py27_0    conda-forge
widgetsnbextension        1.2.6                    py27_3    conda-forge
xlrd                      1.0.0                    py27_0    defaults
xlsxwriter                0.9.6                    py27_0    defaults
xlwings                   0.10.1                   py27_0    defaults
xlwt                      1.1.2                    py27_0    defaults
xz                        5.2.2                         0    conda-forge
yaml                      0.1.6                         0    conda-forge
zeromq                    4.1.5                         0    conda-forge
zict                      0.1.0                    py27_0    conda-forge
zlib                      1.2.8                         3    conda-forge
```

**Comment by shadowwalkersb** [comment]
> OSError: [Errno 2] No such file or directory: './szip-feedstock/recipe/bld.bat'

The last line of the traceback in OP complains about non-existing file that is supposed to be copied. So, it doesn't exist, unless it is coming from a step after the file is copied.

**Comment by jakirkham** [comment]
It may be that the directory name for the recipe needs to match the package name. Is that true of your example?

**Comment by shadowwalkersb** [comment]
See https://github.com/conda-forge/conda-smithy/commit/b4f28fff7a81c91023e75987d1bffd1ee658975b#commitcomment-20951014 and https://github.com/conda-forge/conda-smithy/commit/b4f28fff7a81c91023e75987d1bffd1ee658975b#commitcomment-20951026.

**Comment by jakirkham** [comment]
Could you please share some kind of reproducer with steps?

**Comment by shadowwalkersb** [comment]
Suppose I have a directory `workspace` which is a git repository, so it contains `.git`. I want to create a new feedstock from a recipe `szip-feedstock-location/recipe`. I run `conda smithy init szip-feedstock-location/recipe` in `workspace` and I get the traceback in OP. Then, I added some print statements in conda-smithy. It looks like in `feedstock_io.copy_file`, `repo = get_repo(dst)` the repo returned has a path that looks like `......./workspace/.git`. And there is no `.git` under `szip-feedstock`.

Note also that the exceptions message says
`OSError: [Errno 2] No such file or directory: './szip-feedstock/recipe/bld.bat'` ,
but the file is there. Can't, exactly, follow the code, but it may be that this file is checked if it is in index. And it is not in the index of the upper-level git, the current directory, on the other hand, doesn't even have a git repo, yet.

**Comment by jakirkham** [comment]
As noted in the comments in the commit, it is not supported to create a feedstock repo inside of an existing directory. If you follow this advice and still have an issue, please let me know. As of now I'm considering this a `won't fix`.

**Comment by shadowwalkersb** [comment]
> it is not supported to create a feedstock repo inside of an existing directory

Can you clarify "inside of an existing directory"? How do we create anything inside of a non-existing directory?

**Comment by shadowwalkersb** [comment]
Should also add that I have had this file  organization long before starting to use conda-smithy. And, I am pretty sure it worked before. I even submitted some short instructions to conda-smithy's README. So, I think it used to work, but some changes may have broken it. And, with the current situation it sounds like I need to remove all of my repositories that I have at higher levels to be able to use conda-smithy. Don't really expect you to change things just because my workflow is broken. Just laying out some arguments.

**Comment by shadowwalkersb** [comment]
Feel free to close if there is nothing else to add to the discussion.My take-away is not to use `conda-smithy init` and just initialize feedstock repos manually for my projects.

**Comment by jakirkham** [comment]
> Can you clarify "inside of an existing directory"?

Please see the edit above.

**Comment by jakirkham** [comment]
FWIW I expect PR ( https://github.com/conda-forge/conda-smithy/pull/460 ) will help you to some extent.

A full fix would have to ensure there is always a git repo in advance while handling the fact that a new git repo may need to be created each time. IMHO that is way to much work for me to take on or review at this stage. Plus it might annoy anyone wanting no repo as they would have to keep specifying it. A much better fix would be to deprecate and remove the no repo use case entirely. ( https://github.com/conda-forge/conda-smithy/issues/333 ) While I'd like to do that, I remain unclear on whether anyone is using that functionality or not. Perhaps that is an indication no one is?

**Comment by shadowwalkersb** [comment]
> However, creating a full blown repo inside of another repo can result in confusion and problems down the line. Not to mention this will result in commits with binary blobs of this repo inside of the other repo for any change that occurs (including git gc). This will also lead to performance issues down the line.

I may not be a git expert and I don't claim that my git setup is ideal. But, AFAIK nested repos that are not submodules or subtrees are simply ignored by the parent repo. So, I don't believe nested repos are that dangerous or dangerous at all. At least, I haven't had any problems with my setup.

Discouraging such nested git repos, in a way, sounds like, you can't have your bashrc types of files under version control, because your whole user account would be a git repo and everything else under it would be a nested repo. Another use case is keeping IDE settings under version control. For example, Eclipse has the concept of workspaces where project data and settings are saved. Well, all those projects may be git repos. Then, the suggestion is that I can't keep my eclipse under version control unless my other projects are made submodules of my settings repo. Just doesn't sound natural.

Although, my use cases are, pretty legit if you ask me, 😉 , I'm not asking for any unnecessary efforts put into conda-smithy development. Will take a look at PR #460 and thanks for putting it together!
